monitoring/opencensus/main.go

```diff
@@ -0,0 +1,127 @@
+// Copyright 2021 Google LLC
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     https://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package main
+
+// [START monitoring_sli_metrics_opencensus_setup]
+import (
+	"context"
+	"flag"
+	"fmt"
+	"log"
+	"net/http"
+	"strconv"
+	"time"
+
+	"contrib.go.opencensus.io/exporter/stackdriver"
+	"go.opencensus.io/stats"
+	"go.opencensus.io/stats/view"
+	"go.opencensus.io/tag"
+	"golang.org/x/exp/rand"
+)
+
+// [END monitoring_sli_metrics_opencensus_setup]
+// [START monitoring_sli_metrics_opencensus_measure]
+// set up metrics
+var (
+	requestCount       = stats.Int64("oc_request_count", "total request count", "requests")
+	failedRequestCount = stats.Int64("oc_failed_request_count", "count of failed requests", "requests")
+	responseLatency    = stats.Float64("oc_latency_distribution", "distribution of response latencies", "s")
+)
+
+// [END monitoring_sli_metrics_opencensus_measure]
+
+// set up views
+// [START monitoring_sli_metrics_opencensus_view]
+var (
+	requestCountView = &view.View{
+		Name:        "oc_request_count",
+		Measure:     requestCount,
+		Description: "total request count",
+		Aggregation: view.Count(),
+	}
+	failedRequestCountView = &view.View{
+		Name:        "oc_failed_request_count",
+		Measure:     failedRequestCount,
+		Description: "count of failed requests",
+		Aggregation: view.Count(),
+	}
+	responseLatencyView = &view.View{
+		Name:        "oc_response_latency",
+		Measure:     responseLatency,
+		Description: "The distribution of the latencies",
+		// bucket definitions must be explicit
+		Aggregation: view.Distribution(0, 1000, 2000, 3000, 4000, 5000, 6000, 7000, 8000, 9000, 10000),
+	}
+)
+
+// [END monitoring_sli_metrics_opencensus_view]
+
+func main() {
+
+	// set up project ID
+	projectID := flag.String("project_id", "", "Cloud Project ID")
+	flag.Parse()
+
+	// register the views
+	if err := view.Register(requestCountView, failedRequestCountView, responseLatencyView); err != nil {
+		log.Fatalf("Failed to register the views: %v", err)
+	}
+	// [START monitoring_sli_metrics_opencensus_exporter]
```

Comment:
In general, we try to make sample code work if it were copied and pasted directly from the site into an editor. However, this wouldn't work because it has to be inside `func main`.

Reply:
Do you have a recommendation for how I might proceed...?

Reply:
I guess it depends on how these samples will be shown in the documentation. This fits somewhere in between a quickstart and a snippet, as described by https://github.com/GoogleCloudPlatform/golang-samples/blob/master/CONTRIBUTING.md#style. So, in that sense, there isn't a good guideline to follow.

This works as-is, but I'm concerned about the selective inclusion/exclusion of code -- users won't see everything they need to be able to run the code.

If they take `monitoring_sli_metrics_opencensus_exporter` and copy it into a `.go` file, they will get a compilation error.

Reply:
OK - I get that.  Would adding a comment that this needs to be inside a `func main()` block address your concern? Or would you rather see this perhaps pulled out into a dedicated function? As it stands, it follows the [example](https://opencensus.io/quickstart/go/metrics/#create-the-exporter) almost exactly.

Reply:
No, I don't think that comment is needed. Just something to think about as you're writing the sample and embedding it on a page -- it should be clear to users how to use the code.

Reply:
> No, I don't think that comment is needed. Just something to think about as you're writing the sample and embedding it on a page -- it should be clear to users how to use the code.

Yes, the page (which is [live](https://cloud.google.com/stackdriver/docs/solutions/slo-monitoring/sli-metrics/overview?hl=en) now!) has detailed instructions on how to use the code.

> monitoring_sli_metrics_opencensus_view covers two different scopes (top-level and inside func main). I think that should be avoided as it will look like those pieces of code are right next to each other when embedded.

Yes, that's right.  It's actually fine to have those pieces of code next to each other in a general context.  The views COULD be declared within `func main()`, I just didn't think it would make for elegant code.